File: Loenn/entities/gravitySwitch.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")

local placementData = helpers.createPlacementData('1', {
    defaultToController = true,
    gravityType = consts.gravityTypes.toggle.index,
    cooldown = 1.0,
    textureDirectory = "",
})

local gravitySwitch = {
    name = "GravityHelper/GravitySwitch",
    depth = 2000,
    ignoredFields = consts.ignoredFields,
    fieldInformation = {
        gravityType = consts.fieldInformation.gravityType(0,1,2),
    },
    placements = {
        {
            name = "normal",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.normal.index,
            }),
        },
        {
            name = "inverted",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.inverted.index,
            }),
        },
        {
            name = "toggle",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.toggle.index,
            }),
        },
    },
}

function gravitySwitch.texture(room, entity)
    local basePath = "objects/GravityHelper/gravitySwitch/"
    if entity.textureDirectory and entity.textureDirectory ~= "" then
        basePath = entity.textureDirectory
    end
    basePath = helpers.ensureSingleTrailingSlash(basePath)
    local type = consts.gravityTypeForIndex(entity.gravityType)
    return basePath..(type.switchTexture)
end

function gravitySwitch.selection(room, entity)
    local w,h = 16,16
    return utils.rectangle(entity.x - w/2, entity.y - h/2 + 1, w, h)
end

return gravitySwitch

File: Ahorn/entities/gravitySwitch.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperGravitySwitch

using ..Ahorn, Maple

const PLUGIN_VERSION = "1"

const default_cooldown = 1.0

@mapdef Entity "GravityHelper/GravitySwitch" GravitySwitch(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    defaultToController::Bool=true,
    gravityType::Integer=2,
    cooldown::Real=default_cooldown,
    switchOnHoldables::Bool=true
)

const placements = Ahorn.PlacementDict(
    "Gravity Switch (Normal) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySwitch,
        "point",
        Dict{String, Any}(
            "gravityType" => 0,
        )
    ),
    "Gravity Switch (Inverted) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySwitch,
        "point",
        Dict{String, Any}(
            "gravityType" => 1,
        )
    ),
    "Gravity Switch (Toggle) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySwitch,
        "point",
        Dict{String, Any}(
            "gravityType" => 2,
        )
    ),
)

const gravityTypes = Dict{String, Integer}(
    "Normal" => 0,
    "Inverted" => 1,
    "Toggle" => 2,
)

Ahorn.editingOptions(entity::GravitySwitch) = Dict{String, Any}( "gravityType" => gravityTypes )

const normalSprite = "objects/GravityHelper/gravitySwitch/switch12"
const invertedSprite = "objects/GravityHelper/gravitySwitch/switch01"
const toggleSprite = "objects/GravityHelper/gravitySwitch/toggle01"

Ahorn.editingIgnored(entity::GravitySwitch, multiple::Bool=false) = multiple ? String["x", "y", "modVersion", "pluginVersion"] : String["modVersion", "pluginVersion"]

function Ahorn.selection(entity::GravitySwitch)
    x, y = Ahorn.position(entity)
    return Ahorn.getSpriteRectangle(normalSprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravitySwitch, room::Maple.Room)
    gravityType = get(entity.data, "gravityType", 0)
    sprite = gravityType == 0 ? normalSprite : gravityType == 1 ? invertedSprite : toggleSprite
    return Ahorn.drawSprite(ctx, sprite, 0, 0)
end

end